numpy/core/src/multiarray/item_selection.c

```diff
@@ -2145,11 +2152,15 @@ PyArray_CountNonzero(PyArrayObject *self)
             if (nonzero(data, self)) {
                 ++nonzero_count;
             }
+            if (needs_api && PyErr_Occurred()) {
+                goto finish;
+            }
```

Comment:
I think this should be `needs_api = NpyIter_IterationNeedsAPI(iter)`, for consistency with `NPY_BEGIN_THREADS_NDITER`